### PR TITLE
Split `SearchResult` components into separate files

### PR DESCRIPTION
**Pull request by oisincoveney**
### Description

Splits the `SearchResult` components that were in one file into a separate folder with the components split into separate files. This will help with the [Search Results Facelift](https://github.com/metabase/metabase/pull/34410) PR to ensure that code is both split into relatively logical sections and allows us to make changes to separate sections of the `SearchResult` component more easily.

### How to verify

Tests should still pass - no code changes have been made beyond imports and adding a type here and there.

### CODE REVIEW

File: frontend/src/metabase/search/components/SearchResult/components/Context.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// I think it's very likely that this is a dead codepath: RL 2023-06-21
```
Comment (oisincoveney) [inline frontend/src/metabase/search/components/SearchResult/components/Context.tsx:1]: We might be able to remove this component, but I don't want to touch it in this PR. I'll make changes in the facelift PR if it's deemed to be dead

File: frontend/src/metabase/search/components/SearchResult/components/CollectionIcon.tsx [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+export function CollectionIcon({ item }: { item: WrappedResult }) {
+  const iconProps = { ...item.getIcon() };
+  const isRegular = PLUGIN_COLLECTIONS.isRegularCollection(item.collection);
+  if (isRegular) {
+    iconProps.size = DEFAULT_ICON_SIZE;
+  } else {
+    iconProps.width = 20;
+    iconProps.height = 24;
+  }
+  return <Icon {...iconProps} tooltip={null} />;
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/components/SearchResult/components/CollectionIcon.tsx:15]: I know this code is just moved, but it does look confusing.

```suggestion
  const iconProps = { ...item.getIcon(), tooltip: null };
  const isRegular = PLUGIN_COLLECTIONS.isRegularCollection(item.collection);

  if (isRegular) {
    return <Icon {...iconProps} size={DEFAULT_ICON_SIZE} />;
  }

  return <Icon {...iconProps} width={20} height={24} />;
```

File: frontend/src/metabase/search/components/SearchResult/SearchResult.tsx [hunk 3]
```diff
@@ -1,109 +1,29 @@
+import {
+  ItemIcon,
+  Context,
+  Score,
+} from "metabase/search/components/SearchResult/components";
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/components/SearchResult/SearchResult.tsx:5]:
```suggestion
} from "./components";
```

File: frontend/src/metabase/search/components/SearchResult/components/Context.styled.tsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import styled from "@emotion/styled";
+import { color } from "metabase/lib/colors";
+
+export const ContextText = styled("p")`
+  line-height: 1.4em;
+  color: ${color("text-medium")};
+  margin-top: 0;
+`;
+export const ContextContainer = styled.div`
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/components/SearchResult/components/Context.styled.tsx:9]:
```suggestion
`;

export const ContextContainer = styled.div`
```

File: frontend/src/metabase/search/components/SearchResult/components/Context.styled.tsx [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import styled from "@emotion/styled";
+import { color } from "metabase/lib/colors";
+
+export const ContextText = styled("p")`
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/components/SearchResult/components/Context.styled.tsx:4]: `ContextText` & `ContextContainer` should be removed from `SearchResult.styled.tsx`

File: frontend/src/metabase/search/components/SearchResult/components/Context.tsx [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import { color } from "metabase/lib/colors";
+import {
+  ContextContainer,
+  ContextText,
+} from "metabase/search/components/SearchResult/components/Context.styled";
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/components/SearchResult/components/Context.tsx:6]:
```suggestion
} from "./Context.styled";
```

File: frontend/src/metabase/search/components/SearchResult/components/DefaultIcon.tsx [hunk 7]
```diff
@@ -0,0 +1,7 @@
+import { Icon } from "metabase/core/components/Icon";
+import { DEFAULT_ICON_SIZE } from "metabase/search/components/SearchResult/components";
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/components/SearchResult/components/DefaultIcon.tsx:2]: Circular dependency

```suggestion
import { DEFAULT_ICON_SIZE } from "./icon-size";
```

File: frontend/src/metabase/search/components/SearchResult/components/ItemIcon.styled.tsx [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  if (!active) {
+    return color("text-medium");
+  }
+  if (type === "collection") {
+    return lighten("brand", 0.35);
+  }
+  return color("brand");
+}
+
+export const IconWrapper = styled.div<{
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/components/SearchResult/components/ItemIcon.styled.tsx:21]: `getColorForIconWrapper` & `IconWrapper` should be removed from `SearchResult.styled.tsx`

File: frontend/src/metabase/search/components/SearchResult/components/ItemIcon.tsx [hunk 9]
```diff
@@ -0,0 +1,32 @@
+import type { SearchModelType } from "metabase-types/api";
+import { IconWrapper } from "metabase/search/components/SearchResult/SearchResult.styled";
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/components/SearchResult/components/ItemIcon.tsx:2]:
```suggestion
import { IconWrapper } from "./ItemIcon.styled";
```

File: frontend/src/metabase/search/containers/SearchResultSection.tsx [hunk 10]
```diff
@@ -0,0 +1,11 @@
+import type { WrappedResult } from "metabase/search/types";
+import Card from "metabase/components/Card";
+import { SearchResult } from "metabase/search/components/SearchResult/SearchResult";
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/containers/SearchResultSection.tsx:3]:
```suggestion
import { SearchResult } from "metabase/search/components/SearchResult";
```

**Review by kamilmielnik — CHANGES_REQUESTED** [review]
Please clean up the code that was meant to be moved and make imports consistent.

BTW automatic imports handling in IDEs does not always work as expected - please always verify imports manually.

File: frontend/src/metabase/search/components/SearchResult/components/ItemIcon.tsx [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  type: SearchModelType;
+  active: boolean;
+}) {
+  const IconComponent =
+    type in Object.keys(ModelIconComponentMap)
+      ? ModelIconComponentMap[type as keyof typeof ModelIconComponentMap]
+      : DefaultIcon;
+
+  return (
+    <IconWrapper item={item} type={type} active={active}>
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/components/SearchResult/components/ItemIcon.tsx:28]: There is no `item` prop on `IconWrapper`

File: frontend/src/metabase/nav/components/search/SearchResults/SearchResults.tsx [hunk 12]
```diff
@@ -3,7 +3,7 @@ import { t } from "ttag";
 import { push } from "react-router-redux";
 import { useDebounce } from "react-use";
 import { useListKeyboardNavigation } from "metabase/hooks/use-list-keyboard-navigation";
-import { SearchResult } from "metabase/search/components/SearchResult";
+import { SearchResult } from "metabase/search/components/SearchResult/SearchResult";
```
[the remark this answers is not in the document]
Comment (oisincoveney) [reply frontend/src/metabase/nav/components/search/SearchResults/SearchResults.tsx:6]: I gotta see if there's a Webstorm setting that's causing this, I keep getting caught with this

File: frontend/src/metabase/search/components/SearchResult/components/ItemIcon.tsx [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+      <IconComponent item={item} />
+    </IconWrapper>
+  );
+}
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/components/SearchResult/components/ItemIcon.tsx:32]: WDYT about this?

```suggestion
import type { SearchModelType } from "metabase-types/api";
import { Icon } from "metabase/core/components/Icon";
import type { WrappedResult } from "metabase/search/types";
import { CollectionIcon } from "./CollectionIcon";
import { DefaultIcon } from "./DefaultIcon";
import { IconWrapper } from "./ItemIcon.styled";

interface IconComponentProps {
  item: WrappedResult;
  type: SearchModelType;
}

const IconComponent = ({ item, type }: IconComponentProps) => {
  if (type === "table") {
    return <Icon name="database" />;
  }

  if (type === "collection") {
    return <CollectionIcon item={item} />;
  }

  return <DefaultIcon item={item} />;
};

interface ItemIconProps {
  active: boolean;
  item: WrappedResult;
  type: SearchModelType;
}

export function ItemIcon({ active, item, type }: ItemIconProps) {
  return (
    <IconWrapper type={type} active={active}>
      <IconComponent item={item} type={type} />
    </IconWrapper>
  );
}
```

This way we can:
- get rid of `TableIcon` component
- make sure `item` prop is passed only to the components that need it (previously `TableIcon` received it)
- avoid creating a "framework" for the icons

File: frontend/src/metabase/search/components/SearchResult/components/TableIcon.tsx [hunk 14]
```diff
@@ -0,0 +1,5 @@
+import { Icon } from "metabase/core/components/Icon";
+
+export function TableIcon() {
+  return <Icon name="database" />;
+}
```
[the remark this answers is not in the document]
Comment (kamilmielnik) [reply frontend/src/metabase/search/components/SearchResult/components/TableIcon.tsx:5]: This suggestion can help remove this component: https://github.com/metabase/metabase/pull/34466/files#r1351607495

File: frontend/src/metabase/search/components/SearchResult/components/ItemIcon.tsx (hunk 11 above)
Comment (oisincoveney) [reply frontend/src/metabase/search/components/SearchResult/components/ItemIcon.tsx:28]: Yeah I removed it from the styled component since it just wasn't used anyway, so just needs to be removed here

File: frontend/src/metabase/search/containers/SearchResultSection.tsx [hunk 15]
```diff
@@ -0,0 +1,11 @@
+import type { WrappedResult } from "metabase/search/types";
+import Card from "metabase/components/Card";
+import { SearchResult } from "metabase/search/components/SearchResult";
+
+export const SearchResultSection = ({ items }: { items: WrappedResult[] }) => (
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/containers/SearchResultSection.tsx:5]: Looks like this one belongs in `components`, not `containers`.

File: frontend/src/metabase/search/components/SearchResult/components/ItemIcon.styled.tsx [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+  border: ${({ theme }) => `1px solid ${theme.colors.border[0]}`};
+  border-radius: ${({ theme }) => theme.radius.sm};
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  width: 32px;
+  height: 32px;
+  color: ${({ active, type }) => getColorForIconWrapper({ active, type })};
+  //margin-right: 10px;
+  flex-shrink: 0;
```
Comment (kamilmielnik) [inline frontend/src/metabase/search/components/SearchResult/components/ItemIcon.styled.tsx:34]: Are these changes intentional?
- `border` & `border-radius` were not present before
- `margin-right` was not commented out

**Review by kamilmielnik — APPROVED** [review]
:+1:

File: frontend/src/metabase/search/containers/SearchResultSection.tsx (hunk 15 above)
Comment (oisincoveney) [reply frontend/src/metabase/search/containers/SearchResultSection.tsx:5]: I would say this is more `container` than `component` since it really just acts as a wrapper around the list of search results in `SearchApp`

Comment (oisincoveney) [reply frontend/src/metabase/search/containers/SearchResultSection.tsx:5]: I'll need to take a better look at the actual structure of the `search` folder in general in the future

File: frontend/src/metabase/search/components/SearchResult/components/ItemIcon.styled.tsx (hunk 16 above)
Comment (oisincoveney) [reply frontend/src/metabase/search/components/SearchResult/components/ItemIcon.styled.tsx:34]: Looks like a copy of my current facelift PR - those changes should be present here. Reverted

File: frontend/src/metabase/search/containers/SearchResultSection.tsx (hunk 15 above)
Comment (kamilmielnik) [reply frontend/src/metabase/search/containers/SearchResultSection.tsx:5]: Oh, I though containers are redux-connected components.
